src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/HtmlFormattingPass.cs

```diff
@@ -0,0 +1,105 @@
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.Extensions.Logging;
+using OmniSharp.Extensions.LanguageServer.Protocol.Server;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Formatting
+{
+    internal class HtmlFormattingPass : FormattingPassBase
+    {
+        private readonly ILogger _logger;
+
+        public HtmlFormattingPass(
+            RazorDocumentMappingService documentMappingService,
+            FilePathNormalizer filePathNormalizer,
+            ILanguageServer server,
+            ILoggerFactory loggerFactory)
+            : base(documentMappingService, filePathNormalizer, server)
+        {
+            if (loggerFactory is null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _logger = loggerFactory.CreateLogger<HtmlFormattingPass>();
+        }
+
+        // We want this to run first because it uses the client HTML formatter.
+        public override int Order => DefaultOrder - 5;
+
+        public async override Task<FormattingResult> ExecuteAsync(FormattingContext context, FormattingResult result)
+        {
+            if (context.IsFormatOnType)
+            {
+                // We don't want to handle OnTypeFormatting here.
+                return result;
+            }
+
+            var originalText = context.CodeDocument.GetSourceText();
+
+            var htmlEdits = await HtmlFormatter.FormatAsync(context.CodeDocument, context.Range, context.Uri, context.Options);
+            var normalizedEdits = NormalizeTextEdits(originalText, htmlEdits);
+            var mappedEdits = RemapTextEdits(context.CodeDocument, normalizedEdits, RazorLanguageKind.Html);
+            var changes = mappedEdits.Select(e => e.AsTextChange(originalText));
+            var changedText = originalText.WithChanges(changes);
+
+            TrackEncompassingChange(originalText, changedText, out var spanBeforeChange, out var spanAfterChange);
+            var rangeBeforeEdit = spanBeforeChange.AsRange(originalText);
+            var rangeAfterEdit = spanAfterChange.AsRange(changedText);
+
+            // Create a new formatting context for the changed razor document.
+            var changedContext = await context.WithTextAsync(changedText);
+
+            var indentationChanges = AdjustRazorIndentation(changedContext, (int)rangeAfterEdit.Start.Line, (int)rangeAfterEdit.End.Line);
+            if (indentationChanges.Count > 0)
+            {
+                // Apply the edits that adjust indentation.
+                changedText = changedText.WithChanges(indentationChanges);
+            }
+
+            var finalChanges = SourceTextDiffer.GetMinimalTextChanges(originalText, changedText, lineDiffOnly: false);
```

Comment:
Seems like this is also repeated. Wonder if it's worth just making this part of the orchestration

Reply:
Same answer as https://github.com/dotnet/aspnetcore-tooling/pull/2445#discussion_r479554815.